pkg/app/server.go

```diff
@@ -66,10 +70,20 @@ func (pl promLogger) Log(v ...interface{}) error {
 	return nil
 }
 
-// RunKubeStateMetrics will build and run the kube-state-metrics.
+// RunKubeStateMetricsWrapper runs KSM with context cancellation.
+func RunKubeStateMetricsWrapper(ctx context.Context, opts *options.Options, factories ...customresource.RegistryFactory) error {
+	err := runKubeStateMetrics(ctx, opts, factories...)
+	if ctx.Err() == context.Canceled {
+		klog.Infoln("RESTARTING kube-state-metrics, metrics will be reset")
+		return nil
+	}
+	return err
+}
+
+// runKubeStateMetrics will build and run the kube-state-metrics.
 // Any out-of-tree custom resource metrics could be registered by newing a registry factory
 // which implements customresource.RegistryFactory and pass all factories into this function.
-func RunKubeStateMetrics(ctx context.Context, opts *options.Options, factories ...customresource.RegistryFactory) error {
+func runKubeStateMetrics(ctx context.Context, opts *options.Options, factories ...customresource.RegistryFactory) error {
```

Comment:
KSM is often used as a library by external projects. It's better to keep this method exporter as it was.